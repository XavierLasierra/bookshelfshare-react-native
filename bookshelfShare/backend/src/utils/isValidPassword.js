const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
