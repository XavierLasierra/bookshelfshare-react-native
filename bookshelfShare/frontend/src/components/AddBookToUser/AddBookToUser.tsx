import React, { useEffect, useState } from 'react';
import { View, Text, ActivityIndicator } from 'react-native';
import {
  Menu, MenuOptions, MenuTrigger, MenuOption, renderers
} from 'react-native-popup-menu';
import { useDispatch, useSelector } from 'react-redux';
import { updateUserBooks } from '../../redux/actions/loggedUser.creator';
import { addToShelf } from '../../redux/actions/userShelves.creator';
import stylesConstants from '../../styles/styles.constants';

import styles from './addBookToUser.styles';

export default function AddBookToUser({
  bookIsbn, token, refreshToken, userId, navigation, logo
}: any) {
  const dispatch = useDispatch();
  const books = useSelector((store: any) => store.userBooks);
  const shelves = useSelector((store: any) => store.userShelves);
  const [markListName, setMarkListName] = useState('Mark as');
  const [markShelfName, setMarkShelfName] = useState('Add to shelf');
  const [deleteFrom, setDeleteFrom] = useState('');
  const [deleteFromShelf, setDeleteFromShelf] = useState('');
  const [shelfLocation, setShelfLocation] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  function findInUserBooks() {
    if (books.read.includes(bookIsbn)) {
      setDeleteFrom('read');
      return 'Read';
    }
    if (books.reading.includes(bookIsbn)) {
      setDeleteFrom('reading');
      return 'Reading';
    }
    if (books.wishlist.includes(bookIsbn)) {
      setDeleteFrom('wishlist');
      return 'Wishlist';
    }
    if (books.toRead.includes(bookIsbn)) {
      setDeleteFrom('toRead');
      return 'To Read';
    }
    return 'Mark as';
  }

  function findInUserShelves() {
    let name = 'Add to shelf';
    setDeleteFromShelf('');
    setShelfLocation(null);
    shelves.forEach((shelf: any) => {
      const foundBook = shelf.books
        .find(({ bookIsbn: shelfBookIsbn }: any) => shelfBookIsbn === bookIsbn);
      if (foundBook) {
        // eslint-disable-next-line no-underscore-dangle
        setDeleteFromShelf(shelf._id);
        setShelfLocation(foundBook.customInformation?.location);
        name = shelf.name;
      }
    });
    return name;
  }

  useEffect(() => {
    setMarkListName(findInUserBooks());
    setIsLoading(false);
  }, [books]);

  useEffect(() => {
    setMarkShelfName(findInUserShelves());
  }, [shelves]);

  function handleCategorySelect(value: string) {
    setIsLoading(true);
    dispatch(updateUserBooks(
      userId, {
        bookIsbn,
        deleteFrom,
        addTo: value
      },
      token,
      refreshToken
    ));
  }

  function handleShelfSelect(value: string) {
    if (value === 'delete') {
      dispatch(addToShelf(
        deleteFromShelf,
        // eslint-disable-next-line no-underscore-dangle
        '',
        bookIsbn,
        '',
        token,
        refreshToken
      ));
    } else {
      navigation.push('AddToShelf',
        {
          shelf: shelves.find(({ _id }: any) => _id === value),
          deleteFromShelf,
          bookIsbn,
          logo
        });
    }
  }

  const menus = (
    <>
      <Menu
        onSelect={handleCategorySelect}
        renderer={renderers.NotAnimatedContextMenu}
      >
        <MenuTrigger>
          {isLoading
            ? (
              <View style={styles.activityIndicatorContainer}>
                <ActivityIndicator size="small" color={stylesConstants.colors.dark} />
              </View>
            )
            : <Text style={styles.markButton}>{markListName}</Text>}
        </MenuTrigger>
        <MenuOptions>
          <MenuOption
            value="read"
            text="Read"
            style={styles.menuOption}
          />
          <MenuOption
            value="reading"
            text="Reading"
            style={styles.menuOption}
          />
          <MenuOption
            value="wishlist"
            text="Wishlist"
            style={styles.menuOption}
          />
          <MenuOption
            value="toRead"
            text="To Read"
            style={styles.menuOption}
          />
          <MenuOption
            value=""
            text="Unmark"
            style={styles.menuOption}
          />
        </MenuOptions>
      </Menu>
      <View>
        <Menu
          onSelect={handleShelfSelect}
          renderer={renderers.NotAnimatedContextMenu}
        >
          <MenuTrigger>
            <Text style={styles.markButton}>{markShelfName}</Text>
          </MenuTrigger>
          <MenuOptions>
            {shelves.map(({ _id, name }: any) => (
              <MenuOption
                key={_id}
                value={_id}
                text={name}
                style={styles.menuOption}
              />
            ))}
            <MenuOption
              value="delete"
              text="Delete from shelf"
              style={styles.menuOption}
            />
          </MenuOptions>
        </Menu>
        {shelfLocation
      && (
      <View style={styles.shelfLocationContainer}>
        <Text style={styles.shelfLocationText}>
          Row:
          {shelfLocation[0] + 1}
        </Text>
        <Text style={styles.shelfLocationText}>
          Col:
          {shelfLocation[1] + 1}
        </Text>
      </View>
      )}
      </View>
    </>
  );

  return (
    <View style={styles.addBookToUserContainer}>
      {bookIsbn !== 'Not found'
        ? menus
        : <Text style={styles.canNotSaveText}>Cannot save a book without isbn</Text>}
    </View>
  );
}
