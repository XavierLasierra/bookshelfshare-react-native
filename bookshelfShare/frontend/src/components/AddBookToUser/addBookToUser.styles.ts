import { StyleSheet } from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  addBookToUserContainer: {
    paddingHorizontal: stylesConstants.padding.small,
    paddingBottom: stylesConstants.padding.mid,
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  canNotSaveText: {
    textAlign: 'center',
    fontSize: stylesConstants.fontSize.small,
    width: '100%'
  }
});
