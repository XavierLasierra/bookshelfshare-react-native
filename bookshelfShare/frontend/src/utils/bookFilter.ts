export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
