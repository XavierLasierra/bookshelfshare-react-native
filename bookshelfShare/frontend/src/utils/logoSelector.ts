import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
