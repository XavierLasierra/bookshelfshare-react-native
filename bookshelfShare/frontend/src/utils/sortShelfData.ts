export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
