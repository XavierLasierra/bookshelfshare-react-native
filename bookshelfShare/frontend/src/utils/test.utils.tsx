import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
