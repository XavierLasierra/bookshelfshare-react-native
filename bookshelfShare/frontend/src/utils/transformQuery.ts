interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
