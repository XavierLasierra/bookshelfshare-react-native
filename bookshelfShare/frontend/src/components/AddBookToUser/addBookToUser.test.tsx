import React from 'react';
import { MenuProvider } from 'react-native-popup-menu';
import AddBookToUser from './AddBookToUser';
import { render } from '../../utils/test.utils';
import userBooksMock from '../../mocks/userBooks.mock';

describe('Given an AddBookToUser component', () => {
  describe('When it is rendered', () => {
    describe('And isbn is Not found', () => {
      test('Then should match the snapshot', () => {
        const screen = render(
          <AddBookToUser bookIsbn="Not found" />
        );
        expect(screen).toMatchSnapshot();
      });
    });

    describe('And isbn is defined', () => {
      test('Then should match the snapshot', () => {
        const initialState = { userBooks: userBooksMock };
        const screen = render(
          <MenuProvider>
            <AddBookToUser bookIsbn="1234" />
          </MenuProvider>,
          initialState
        );
        expect(screen).toMatchSnapshot();
      });
    });
  });
});
