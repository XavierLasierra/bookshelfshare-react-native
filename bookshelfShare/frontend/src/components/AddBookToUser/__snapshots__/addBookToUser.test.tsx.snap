// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an AddBookToUser component When it is rendered And isbn is Not found Then should match the snapshot 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingBottom": 20,
      "paddingHorizontal": 10,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 12,
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    Cannot save a book without isbn
  </Text>
</View>
`;

exports[`Given an AddBookToUser component When it is rendered And isbn is defined Then should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 20,
          "paddingHorizontal": 10,
        }
      }
    >
      <View>
        <View
          collapsable={false}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              ctx={
                Object {
                  "menuActions": Object {
                    "_getOpenedMenu": [Function],
                    "_notify": [Function],
                    "closeMenu": [Function],
                    "isMenuOpen": [Function],
                    "openMenu": [Function],
                    "toggleMenu": [Function],
                  },
                  "menuRegistry": Object {
                    "getAll": [Function],
                    "getMenu": [Function],
                    "setOptionsCustomStyles": [Function],
                    "subscribe": [Function],
                    "unsubscribe": [Function],
                    "updateLayoutInfo": [Function],
                  },
                }
              }
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#ECECEC",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 40,
                    "lineHeight": 40,
                    "textAlign": "center",
                    "width": 150,
                  }
                }
              >
                Mark as
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View>
        <View>
          <View
            collapsable={false}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                ctx={
                  Object {
                    "menuActions": Object {
                      "_getOpenedMenu": [Function],
                      "_notify": [Function],
                      "closeMenu": [Function],
                      "isMenuOpen": [Function],
                      "openMenu": [Function],
                      "toggleMenu": [Function],
                    },
                    "menuRegistry": Object {
                      "getAll": [Function],
                      "getMenu": [Function],
                      "setOptionsCustomStyles": [Function],
                      "subscribe": [Function],
                      "unsubscribe": [Function],
                      "updateLayoutInfo": [Function],
                    },
                  }
                }
                style={
                  Array [
                    undefined,
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "borderColor": "#ECECEC",
                      "borderRadius": 10,
                      "borderWidth": 1,
                      "height": 40,
                      "lineHeight": 40,
                      "textAlign": "center",
                      "width": 150,
                    }
                  }
                >
                  Add to shelf
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
